File: src/main/prim/_of.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.Syntax
import org.nlogo.api.{ I18N, LogoListBuilder }
import org.nlogo.agent.{ Agent, AgentSet }
import org.nlogo.nvm.{ Reporter, Context, EngineException }

class _of extends Reporter {

  override def report(context: Context): AnyRef =
    args(1).report(context) match {
      case agent: Agent =>
        if (agent.id == -1)
          throw new EngineException(
            context, this,
            I18N.errors.getN("org.nlogo.$common.thatAgentIsDead",
                             agent.classDisplayName))
        args(0).checkAgentClass(agent, context)
        new Context(context, agent)
          .evaluateReporter(agent, args(0))
      case sourceSet: AgentSet =>
        val builder = new LogoListBuilder
        val freshContext = new Context(context, sourceSet)
        args(0).checkAgentSetClass(sourceSet, context)
        val iter = sourceSet.shufflerator(context.job.random)
        while(iter.hasNext)
          builder.add(freshContext.evaluateReporter(iter.next(), args(0)))
        builder.toLogoList
      case x =>
        throw new org.nlogo.nvm.ArgumentTypeException(
          context, this, 1, Syntax.AgentsetType | Syntax.AgentType, x)
    }

}

File: netlogo-gui/src/main/prim/_of.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Agent;
import org.nlogo.agent.AgentIterator;
import org.nlogo.agent.AgentSet;
import org.nlogo.core.I18N;
import org.nlogo.api.LogoException;
import org.nlogo.api.LogoListBuilder;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.RuntimePrimitiveException;
import org.nlogo.nvm.Reporter;

public final class _of
    extends Reporter {


  @Override
  public Object report(final Context context) throws LogoException {
    Object agentOrSet = args[1].report(context);
    if (agentOrSet instanceof Agent) {
      Agent agent = (Agent) agentOrSet;
      if (agent.id() == -1) {
        throw new RuntimePrimitiveException(context, this,
          I18N.errorsJ().getN("org.nlogo.$common.thatAgentIsDead", agent.classDisplayName()));
      }
      args[0].checkAgentClass(agent, context);
      return new Context(context, agent).evaluateReporter(agent, args[0]);
    } else if (agentOrSet instanceof AgentSet) {
      AgentSet sourceSet = (AgentSet) agentOrSet;
      LogoListBuilder result = new LogoListBuilder();
      Context freshContext = new Context(context, sourceSet);
      args[0].checkAgentSetClass(sourceSet, context);
      for (AgentIterator iter = sourceSet.shufflerator(context.job.random);
           iter.hasNext();) {
        result.add(freshContext.evaluateReporter(iter.next(), args[0]));
      }
      return result.toLogoList();
    } else {
      throw new org.nlogo.nvm.ArgumentTypeException
          (context, this, 1,
              Syntax.AgentsetType() | Syntax.AgentType(),
              agentOrSet);
    }
  }
}
